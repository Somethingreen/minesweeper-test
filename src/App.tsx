import React from 'react';

import { Container, AppBar, Toolbar, Typography } from '@mui/material';
import { Game } from './features/game/Game';

function App() {
  return (
    <React.Fragment>
      <AppBar>
        <Toolbar>
          <Typography variant="h6">
            Minesweeper
          </Typography>
        </Toolbar>
      </AppBar>
      <Toolbar />
      <Game />
    </React.Fragment>
  );
}

export default App;
