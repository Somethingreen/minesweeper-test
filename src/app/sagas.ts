import { EventChannel } from 'redux-saga';
import { take, put, apply } from 'redux-saga/effects';
import { PayloadAction } from '@reduxjs/toolkit';


export function* socketSend(ws: WebSocket, action: PayloadAction<string>) {
  yield apply(ws, 'send', [action.payload]);
}

export function* watchSocketChannel(chan: EventChannel<unknown>): any {
  while (true) {
    try {
      const action = yield take(chan);
      yield put(action);
    } catch (err) {
      console.error(err);
    }
  }
}
