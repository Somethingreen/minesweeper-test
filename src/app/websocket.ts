import { Action, PayloadAction } from "@reduxjs/toolkit";
import { EventChannel, eventChannel } from "redux-saga";

export const actions = {
  SOCKET_OPEN: 'socket/open',
  SOCKET_ERROR: 'socket/error',
  SOCKET_CLOSED: 'socket/closed',
  SOCKET_MESSAGE: 'socket/message',
  SOCKET_SEND: 'socket/send'
}

const socketOpen = (_: Event): Action => ({ type: actions.SOCKET_OPEN });
const socketError = (_: Event): Action => ({ type: actions.SOCKET_ERROR });
const socketClosed = (_: CloseEvent): Action => ({ type: actions.SOCKET_CLOSED });
const socketMessage = (e: MessageEvent): PayloadAction<string> => ({ type: actions.SOCKET_MESSAGE, payload: e.data });

export function createEventChannel(ws: WebSocket): EventChannel<unknown> {
  return eventChannel((emit) => {
    ws.addEventListener('open', (e: Event) => emit(socketOpen(e)));
    ws.addEventListener('error', (e) => emit(socketError(e)));
    ws.addEventListener('close', (e: CloseEvent) => emit(socketClosed(e)));
    ws.addEventListener('message', (e: MessageEvent) => emit(socketMessage(e)));
    return () => ws.close();
  });
}

export function socketSend(message: string): PayloadAction<string> {
    return { type: actions.SOCKET_SEND, payload: message };
}

export function createWebsocketConnection(url: string): WebSocket {
  return new WebSocket(url);
}
