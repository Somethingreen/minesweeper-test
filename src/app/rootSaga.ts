import { call, takeEvery, all } from 'redux-saga/effects';
import { createWebsocketConnection, actions, createEventChannel } from './websocket';
import { WEBSOCKET_URL } from './const';
import { gameSaga } from '../features/game/gameSagas';
import { socketSend, watchSocketChannel } from './sagas';


export default function* rootSaga(): any {
    const ws = createWebsocketConnection(WEBSOCKET_URL);
    const chan = yield call(createEventChannel, ws);

    yield takeEvery(actions.SOCKET_SEND, socketSend, ws);

    yield all([
        gameSaga(),
        watchSocketChannel(chan)
    ]);

}
