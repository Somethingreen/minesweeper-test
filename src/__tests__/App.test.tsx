import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import App from '../App';

it('Renders start menu', () => {
  const { getByText, getByRole, getAllByRole, asFragment } = render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(asFragment).toMatchSnapshot();

  expect(getByText(/Minesweeper/i)).toBeInTheDocument();
  expect(getAllByRole("radio").length).toEqual(4);
  const startMenuButton = getByRole("button", { name: /Start/i });
  expect(startMenuButton).toBeInTheDocument();
});
