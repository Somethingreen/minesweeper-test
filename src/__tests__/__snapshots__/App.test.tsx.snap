// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders start menu 1`] = `[Function]`;
